File: python/usb_gpu_pipeline.py
#!/usr/bin/env python3
"""
apps/python/usb_gpu_pipeline.py

Python version of apps/cpp/usb_gpu_pipeline.cpp.

All camera capture, upload, encoding, and RTSP serving run in C++ threads.
Python only "visits" the frame data to apply Gaussian blur (--modify flag),
using TensorPythonInterface as the pass-through point.

Pipeline (default — passthrough, no Python processing):
    [C++] USBCameraThread → UploadGPUFrameFilter(CUDA)
        → DecodedToTensorFrameFilter(RGB)
        → TensorPythonInterface  ← Python consumer (passes frame straight through)
        → TensorToDecodedFrameFilter(RGB)
        → EncodingFrameFilter(NVENC H264) → RTSPMuxerFrameFilter → RTSPServerThread

Pipeline (--modify — Gaussian blur in Python, mirrors C++ GPUOpenCVThread):
    [C++] USBCameraThread → UploadGPUFrameFilter(CUDA)
        → DecodedToTensorFrameFilter(RGB)
        → TensorPythonInterface  ← Python: pull GPU tensor, blur on CPU, push back
        → TensorToDecodedFrameFilter(RGB)
        → EncodingFrameFilter(NVENC H264) → RTSPMuxerFrameFilter → RTSPServerThread

Usage:
    python3 apps/python/usb_gpu_pipeline.py [options]
    python3 apps/python/usb_gpu_pipeline.py --modify   # enable Gaussian blur

Then connect with:
    ffplay rtsp://localhost:8554/live/stream
    ffplay -rtsp_transport tcp rtsp://localhost:8554/live/stream
    ffplay -fflags nobuffer -flags low_delay -framedrop -probesize 32 -analyzeduration 0 rtsp://localhost:8554/live/stream

Press Ctrl+C to stop.
"""

import sys
import time
import argparse
import threading

import limef

try:
    import torch
    import torch.nn.functional as F_torch
    _TORCH = True
except ImportError:
    _TORCH = False


def _make_gauss_kernel(device):
    """Build a 15×15 Gaussian kernel matching OpenCV GaussianBlur(15,15,0)."""
    ksize = 15
    # OpenCV sigma formula for ksize=15: 0.3*((15-1)*0.5 - 1) + 0.8 = 2.6
    sigma = 0.3 * ((ksize - 1) * 0.5 - 1) + 0.8
    x = torch.arange(ksize, dtype=torch.float32, device=device) - ksize // 2
    gauss = torch.exp(-x.pow(2.0) / (2.0 * sigma ** 2))
    gauss = gauss / gauss.sum()
    kernel_2d = gauss.outer(gauss)                            # (15, 15)
    # Shape for grouped conv: (out_channels, in_channels/groups, H, W)
    # groups=C means one filter per channel
    return kernel_2d.view(1, 1, ksize, ksize)


def main():
    p = argparse.ArgumentParser(
        description='limef USB camera → GPU → RTSP pipeline with optional Python frame processing',
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    p.add_argument('-m', '--modify',  action='store_true',
                   help='enable Gaussian blur in Python (mirrors C++ GPUOpenCVThread)')
    p.add_argument('-d', '--device',  default='/dev/video0', metavar='DEV',
                   help='V4L2 camera device')
    p.add_argument('-p', '--port',    type=int, default=8554,
                   help='RTSP server port')
    p.add_argument('-w', '--width',   type=int, default=640)
    p.add_argument('-H', '--height',  type=int, default=480)
    p.add_argument('-f', '--fps',     type=int, default=30)
    p.add_argument('--bitrate',       type=int, default=4_000_000,
                   help='NVENC target bitrate in bits/sec')
    p.add_argument('--url-tail',      default='/live/stream', metavar='PATH',
                   help='RTSP URL path component')
    args = p.parse_args()

    port     = args.port
    url_tail = args.url_tail
    SLOT     = 1
    TIMEOUT_MS = 200

    print("==========================================")
    print("  USB Camera → GPU → RTSP (Python)")
    print("==========================================")
    print(f"Device:     {args.device}")
    print(f"Resolution: {args.width}x{args.height} @ {args.fps} fps")
    print(f"Port:       {port}")
    print(f"URL:        rtsp://localhost:{port}{url_tail}")
    print(f"Modify:     {args.modify}  (Gaussian blur 15×15 in Python)")
    print("==========================================")
    print("Press Ctrl+C to stop\n")

    # ── C++ camera source ──────────────────────────────────────────────────────
    # USBCameraThread: YUYV → internal SwScale → NV12 DecodedFrame
    cam_ctx               = limef.USBCameraContext(args.device, SLOT)
    cam_ctx.width         = args.width
    cam_ctx.height        = args.height
    cam_ctx.fps           = args.fps
    cam_ctx.output_format = limef.AV_PIX_FMT_NV12  # suitable for GPU upload

    camera = limef.USBCameraThread('usb-camera', cam_ctx)

    # ── C++ upstream chain (before Python visit) ───────────────────────────────
    # NV12 DecodedFrame → GPU upload → TensorFrame (3, H, W) RGB on GPU
    upload = limef.UploadGPUFrameFilter('gpu-upload', limef.HWACCEL_CUDA)
    d2t    = limef.DecodedToTensorFrameFilter('d2t', limef.CHANNEL_ORDER_RGB)

    # ── TensorPythonInterface — Python visits here ─────────────────────────────
    # GPU TensorFrames arrive via pull(); Python may process and must push() back.
    #
    # TensorPythonInterface fifo: stack_size=10, leaky=True, hw_accel=CUDA
    #   10 TensorFrame slots absorb a pull() delay in the Python consumer loop.
    #   leaky=True: drop GPU frames if the Python loop falls behind rather than
    #   stalling the camera → upload → d2t chain.
    #   hw_accel=CUDA: CPU TensorFrames are uploaded to the GPU at the thread
    #   boundary (H2D copy) before they become available from pull().
    pyf    = limef.TensorPythonInterface(stack_size=10, leaky=True,
                                         hw_accel=limef.HWACCEL_CUDA, fifo_size=0)
    client = pyf.client()

    # ── C++ downstream chain (after Python visit) ──────────────────────────────
    # TensorFrame (3, H, W) RGB GPU → DecodedFrame CUDA NV12 → NVENC → RTP → RTSP
    t2d = limef.TensorToDecodedFrameFilter('t2d', limef.CHANNEL_ORDER_RGB)

    enc_params              = limef.FFmpegEncoderParams()
    enc_params.codec_id     = limef.AV_CODEC_ID_H264
    enc_params.hw_accel     = limef.HWACCEL_CUDA
    enc_params.bitrate      = args.bitrate
    enc_params.preset       = 'p1'
    enc_params.tune         = 'ull'
    enc_params.max_b_frames = 0
    enc_params.gop_size     = args.fps // 2

    encoder   = limef.EncodingFrameFilter('encoder', enc_params)
    rtp_muxer = limef.RTSPMuxerFrameFilter('rtp-muxer')
    # stack_size=30: absorbs I-frame bursts (several RTP packets in rapid succession).
    # fifo_size=100: cap on queued RTP packets; prevents unbounded growth under slow clients.
    # leaky is always False for RTSPServerThread — RTP sequence gaps corrupt the stream.
    rtsp      = limef.RTSPServerThread('rtsp-server', port=port, stack_size=30, fifo_size=100)

    # ── Wire the pipeline ──────────────────────────────────────────────────────
    # camera → upload → d2t → [Python] → t2d → encoder → rtp → rtsp
    camera.cc(upload).cc(d2t).cc(pyf.getInput())
    pyf.getOutput().cc(t2d).cc(encoder).cc(rtp_muxer).cc(rtsp.getInput())

    # ── Python consumer thread ─────────────────────────────────────────────────
    stop_event  = threading.Event()
    frame_count = [0]
    t_start     = [time.monotonic()]

    # Gaussian kernel is built lazily on the first frame so we get the right device.
    gauss_kernel    = [None]   # [Tensor or None]
    warned_cpu_push = [False]  # one-shot format-mismatch warning

    def consumer():
        while not stop_event.is_set():
            frame = client.pull(timeout_ms=TIMEOUT_MS)

            if frame is None:
                continue

            if isinstance(frame, limef.StreamFrame):
                # Forward stream metadata so EncodingFrameFilter can configure itself
                client.push(frame)
                continue

            if not isinstance(frame, limef.TensorFrame):
                continue

            frame_count[0] += 1

            if args.modify and frame.is_gpu and _TORCH:
                # ── GPU Gaussian blur — mirrors C++ GPUOpenCVThread ─────────────
                # PIPELINE RULE: always push GPU TensorFrames into this pipeline.
                # TensorToDecodedFrameFilter selects its output format based on the
                # incoming frame: GPU → AV_PIX_FMT_CUDA (NV12) → NVENC ✓
                #                 CPU → AV_PIX_FMT_GBRP             → NVENC ✗ (wrong colours)
                # So any processing that produces CPU data must upload back to GPU
                # before pushing. Here we stay on GPU the whole time via torch.
                t = torch.from_dlpack(frame.planes[0]).float()  # (3,H,W) float32 CUDA
                C, H, W = t.shape

                # Lazy-init kernel on the same device as the tensor
                if gauss_kernel[0] is None:
                    gauss_kernel[0] = _make_gauss_kernel(t.device).expand(C, 1, 15, 15).contiguous()

                # Grouped depthwise conv: one filter per channel, padding=7 → same size
                blurred = F_torch.conv2d(
                    t.unsqueeze(0), gauss_kernel[0], padding=7, groups=C
                ).squeeze(0).clamp(0, 255).to(torch.uint8)    # (3,H,W) uint8 CUDA

                # Write into a new owned GPU TensorFrame.
                # reserve_gpu_plane allocates a fresh cudaMalloc buffer.
                # torch.from_dlpack gives a writable view; copy_ fills it.
                # The C++ Plane.d_data_ remains valid after the DLPack capsule
                # is consumed by torch, so t2d can read it downstream.
                out = limef.TensorFrame()
                out.reserve_gpu_plane(0, [C, H, W], 'uint8')
                torch.from_dlpack(out.planes[0]).copy_(blurred)
                out.timestamp = frame.timestamp
                out.slot      = frame.slot
                client.push(out)
            else:
                # No processing (or torch unavailable) — pass frame straight through.
                # Warn once if a CPU frame is about to enter the CUDA pipeline.
                if (isinstance(frame, limef.TensorFrame)
                        and not frame.is_gpu
                        and not warned_cpu_push[0]):
                    print("WARNING: pushing CPU TensorFrame into a CUDA/NVENC pipeline — "
                          "TensorToDecodedFrameFilter will output GBRP instead of "
                          "CUDA NV12; NVENC will produce wrong colours. "
                          "Ensure your processing keeps frames on GPU.",
                          file=sys.stderr)
                    warned_cpu_push[0] = True
                client.push(frame)

            if frame_count[0] % 100 == 1:
                elapsed = time.monotonic() - t_start[0]
                print(f"  frame #{frame_count[0]:5d}"
                      f"  gpu={frame.is_gpu}"
                      f"  ts={frame.timestamp / 1e6:7.3f} s"
                      f"  elapsed={elapsed:.1f} s"
                      f"  fps={frame_count[0] / max(elapsed, 1e-9):.1f}")

    consumer_thread = threading.Thread(target=consumer, daemon=True,
                                       name='limef-tensor-consumer')

    # ── Start (downstream first, then upstream) ────────────────────────────────
    print("Starting RTSP server ...")
    rtsp.start()
    time.sleep(0.1)

    rtsp.expose(SLOT, url_tail)
    time.sleep(0.05)

    print("Starting consumer thread ...")
    consumer_thread.start()

    print("Starting USB camera ...")
    camera.start()
    time.sleep(0.5)

    print(f"\nReady!  Connect with:")
    print(f"  ffplay rtsp://localhost:{port}{url_tail}")
    print(f"  ffplay -rtsp_transport tcp rtsp://localhost:{port}{url_tail}")
    print(f"  ffplay -fflags nobuffer -flags low_delay -framedrop "
          f"-probesize 32 -analyzeduration 0 rtsp://localhost:{port}{url_tail}\n")

    # ── Main loop ──────────────────────────────────────────────────────────────
    try:
        while True:
            time.sleep(0.5)
    except KeyboardInterrupt:
        print("\nShutting down...")

    # ── Cleanup (upstream first, then downstream) ──────────────────────────────
    stop_event.set()

    print("Stopping USB camera ...")
    try:
        camera.stop()
    except KeyboardInterrupt:
        print("Interrupted — forcing exit.")
        sys.exit(1)

    # Wait for the consumer thread to exit before destroying C++ objects.
    # Same rationale as pipeline_test_thread_tensor.py: pull() may be blocking
    # in fifo_.read() with the GIL released; destroying TensorFrameFifo while
    # a thread waits on its condition variable → std::terminate().
    while consumer_thread.is_alive():
        try:
            consumer_thread.join(timeout=TIMEOUT_MS / 1000 + 0.5)
        except KeyboardInterrupt:
            pass

    print("Stopping RTSP server ...")
    try:
        rtsp.stop()
    except KeyboardInterrupt:
        print("Interrupted — forcing exit.")
        sys.exit(1)

    elapsed = time.monotonic() - t_start[0]
    print(f"\nDone.  {elapsed:.1f} s, {frame_count[0]} frames"
          f"  ({frame_count[0] / max(elapsed, 1e-9):.1f} fps avg)")


if __name__ == '__main__':
    main()

File: cpp/opencv/usb_gpu_pipeline.cpp
/*
 * Copyright (c) 2026 Sampsa Riikonen <sampsa.riikonen@iki.fi>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND.
 */
/*
 * usb_gpu_pipeline - Full USB camera to RTSP pipeline via GPU
 *
 * Pipeline (default, no --modify):
 *   USBCameraThread → UploadGPUFrameFilter → EncodingFrameFilter(NVENC) → RTSPMuxerFrameFilter → RTSPServer
 *
 * Pipeline (with --modify, GPU Gaussian blur via OpenCV):
 *   USBCameraThread → UploadGPUFrameFilter → DecodedToTensorFrameFilter
 *       → GPUOpenCVThread (Gaussian blur on TensorFrame)
 *       → TensorToDecodedFrameFilter → EncodingFrameFilter(NVENC) → RTSPMuxerFrameFilter → RTSPServer
 *
 * NOTE: one copy could be eliminated here:
 * DecodedToTensorFrameFilter copies DecodedFrame -> TensorFrame
 * GPUOpenCVThread copies TensorFrame -> TensorFrame
 * We could have a FrameFifo that copies DecodedFrame directly into a TensorFrame taken from the FrameFifo's stack
 * 
 * Usage:
 *   ./usb_gpu_pipeline [--device /dev/video0] [--port 8554] [--width 640] [--height 480]
 *   ./usb_gpu_pipeline --modify    # enable GPU Gaussian blur
 *
 * Then connect with:
 *   ffplay rtsp://localhost:8554/live/stream
 *   ffplay -rtsp_transport tcp rtsp://localhost:8554/live/stream
 *   ffplay -fflags nobuffer -flags low_delay -framedrop -probesize 32 -analyzeduration 0 rtsp://localhost:8554/live/stream
 *
 * Press Ctrl+C to stop.
 *
 * ## Latency analysis (~500 ms end-to-end measured)
 *
 * Stage                          Latency    Notes
 * ─────────────────────────────────────────────────────────────────────
 * USB camera sensor              33–100 ms  Hardware, unavoidable at 30 fps
 * V4L2 capture (1 buffer)           ~0 ms  Single buffer is correct for low latency
 * SwScale YUYV→NV12                 ~1 ms  Synchronous in camera thread
 * GPU upload + NVENC encode        5–15 ms  Fast with preset=p1, tune=ull
 * GOP wait (gop_size=5 @ 30fps)   0–166 ms  IDR interval; larger → more client wait on connect
 * RTSP + network (localhost)         ~5 ms
 * ffplay decode + display         50–200 ms  Largest variable; use -fflags nobuffer -framedrop
 *
 * Key tuning levers:
 *  - enc_params.gop_size: smaller = faster client connect, higher bitrate.
 *    gop_size=1 (all-intra) is maximum responsiveness; gop_size=5-10 is a good tradeoff.
 *  - NVENC preset p1 + tune ull: already at minimum encoder latency.
 *  - V4L2 req.count=1 in usbcamera.cpp: keeps only the freshest frame, no camera-side queue.
 *  - ffplay flags: -fflags nobuffer -flags low_delay -framedrop -probesize 32 -analyzeduration 0
 */

#include <iostream>
#include <string>
#include <atomic>
#include <csignal>
#include <thread>
#include <chrono>
#include <getopt.h>

#include "gpu_opencv_thread.h"
#include "limef/thread/usbcamera.h"
#include "limef/framefilter/uploadgpu.h"
#include "limef/framefilter/decoded_to_tensor.h"
#include "limef/framefilter/tensor_to_decoded.h"
#include "limef/framefilter/encoding.h"
#include "limef/framefilter/rtspmuxer.h"
#include "limef/framefilter/dump.h"
#include "limef/rtsp/rtspserverthread.h"
#include "limef/encode/ffmpeg_encoder.h"

static std::atomic<bool> g_running{true};

static void signal_handler(int signum) {
    if (signum == SIGINT || signum == SIGTERM) {
        std::cout << "\nShutting down..." << std::endl;
        g_running = false;
    }
}

static void print_usage(const char* prog) {
    std::cout << "Usage: " << prog << " [options]\n"
              << "\n"
              << "Options:\n"
              << "  --modify, -m          Demo modify the image by blurring on the GPU\n"
              << "  --device, -d <path>   USB camera device (default: /dev/video0)\n"
              << "  --port,   -p <port>   RTSP port (default: 8554)\n"
              << "  --width,  -w <px>     Capture width (default: 640)\n"
              << "  --height, -H <px>     Capture height (default: 480)\n"
              << "  --fps,    -f <fps>    Capture FPS (default: 30)\n"
              << "  --help,   -h          Show this help\n"
              << "\n"
              << "Example:\n"
              << "  " << prog << " --device /dev/video0 --port 8554\n"
              << "\n"
              << "Then connect with:\n"
              << "  ffplay rtsp://localhost:8554/live/stream\n"
              << std::endl;
}

int main(int argc, char** argv) {
    std::string device = "/dev/video0";
    int port = 8554;
    int width = 640;
    int height = 480;
    int fps = 30;
    bool modify = false;

    static struct option long_options[] = {
        {"modify", no_argument, 0, 'm'},
        {"device", required_argument, 0, 'd'},
        {"port",   required_argument, 0, 'p'},
        {"width",  required_argument, 0, 'w'},
        {"height", required_argument, 0, 'H'},
        {"fps",    required_argument, 0, 'f'},
        {"help",   no_argument,       0, 'h'},
        {0, 0, 0, 0}
    };

    int opt;
    while ((opt = getopt_long(argc, argv, "md:p:w:H:f:h", long_options, nullptr)) != -1) {
        switch (opt) {
            case 'd': device = optarg; break;
            case 'p': port = std::stoi(optarg); break;
            case 'w': width = std::stoi(optarg); break;
            case 'H': height = std::stoi(optarg); break;
            case 'f': fps = std::stoi(optarg); break;
            case 'm': modify = true; break;
            case 'h':
            default:
                print_usage(argv[0]);
                return (opt == 'h') ? 0 : 1;
        }
    }

    std::signal(SIGINT, signal_handler);
    std::signal(SIGTERM, signal_handler);

    using namespace Limef;
    using namespace Limef::thread;
    using namespace Limef::ff;

    const int SLOT = 1;
    const char* URL_TAIL = "/live/stream";

    std::cout << "==========================================\n";
    std::cout << "  USB Camera → GPU TensorFrame → RTSP Pipeline\n";
    std::cout << "==========================================\n";
    std::cout << "Device:     " << device << "\n";
    std::cout << "Resolution: " << width << "x" << height << " @ " << fps << " fps\n";
    std::cout << "Port:       " << port << "\n";
    std::cout << "URL:        rtsp://localhost:" << port << URL_TAIL << "\n";
    std::cout << "==========================================\n";
    std::cout << "Press Ctrl+C to stop\n\n";

    // --- 1. USB Camera (producer) ---
    USBCameraContext cam_ctx(device, SLOT);
    cam_ctx.width = width;
    cam_ctx.height = height;
    cam_ctx.fps = fps;
    cam_ctx.output_format = AV_PIX_FMT_NV12;

    USBCameraThread camera("usb-camera", cam_ctx);

    // --- 2. GPU Upload ---
    UploadGPUParams upload_params(HWAccel::CUDA);
    UploadGPUFrameFilter upload("gpu-upload", upload_params);

    // --- 3. DecodedFrame → TensorFrame (GPU, NV12 → CHW RGB) ---
    DecodedToTensorFrameFilter d2t("d2t", ChannelOrder::RGB);

    // --- 4. GPU OpenCV Processing (TensorFrame in, TensorFrame out) ---
    FrameFifoContext opencv_ctx(false, 5, 0, HWAccel::CUDA, "");
    LimefApp::GPUOpenCVThread opencv("gpu-opencv", opencv_ctx);

    // --- 5. TensorFrame → DecodedFrame (GPU, CHW RGB → NV12 CUDA) ---
    TensorToDecodedFrameFilter t2d("t2d", ChannelOrder::RGB);

    // --- 6. NVENC H.264 Encoding ---
    encode::FFmpegEncoderParams enc_params;
    enc_params.codec_id = AV_CODEC_ID_H264;
    enc_params.hw_accel = HWAccel::CUDA;
    enc_params.bitrate = 4*1024*1024; // 4Mbps
    enc_params.preset = std::string("p1");
    enc_params.tune = std::string("ull");
    enc_params.max_b_frames = 0;
    enc_params.gop_size = fps/2;
    EncodingFrameFilter encoder("nvenc-encoder", enc_params);

    // --- 7. RTP Muxer ---
    RTSPMuxerFrameFilter rtp_muxer("rtp-muxer");

    // --- 8. RTSP Server ---
    FrameFifoContext rtsp_ctx(false, 5, 100);
    Limef::rtsp::RTSPServerThread rtsp_server("rtsp-server", rtsp_ctx, port);

    // --- Debug: dump filters (uncomment wiring below to enable) ---
    DumpFrameFilter dump_after_camera("DUMP-after-camera");
    DumpFrameFilter dump_after_upload("DUMP-after-upload");
    DumpFrameFilter dump_after_opencv("DUMP-after-opencv");
    DumpFrameFilter dump_after_encoder("DUMP-after-encoder");
    DumpFrameFilter dump_after_rtp("DUMP-after-rtp");

    // --- Wire the pipeline ---

    if (modify) {
        // USBCamera → Upload → d2t → GPUOpenCV(TensorFrame) → t2d → Encoder → RTPMuxer → RTSPServer
        camera.getOutput().cc(upload).cc(d2t).cc(opencv.getInput());
        opencv.getOutput().cc(t2d).cc(encoder).cc(rtp_muxer).cc(rtsp_server.getInput());
    }
    else {
        // just encode and transmit
        camera.getOutput().cc(upload).cc(encoder).cc(rtp_muxer).cc(rtsp_server.getInput());
    }

    // Debug version with dumps at each stage:
    // camera.getOutput().cc(dump_after_camera).cc(upload).cc(dump_after_upload).cc(d2t).cc(opencv.getInput());
    // opencv.getOutput().cc(dump_after_opencv).cc(t2d).cc(encoder).cc(dump_after_encoder).cc(rtp_muxer).cc(dump_after_rtp).cc(rtsp_server.getInput());

    // --- Start (reverse order: downstream first) ---
    std::cout << "Starting RTSP server..." << std::endl;
    rtsp_server.start();
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    rtsp_server.expose(SLOT, URL_TAIL);
    std::this_thread::sleep_for(std::chrono::milliseconds(50));

    std::cout << "Starting GPU OpenCV thread..." << std::endl;
    opencv.start();
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "Starting USB camera..." << std::endl;
    camera.start();
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    std::cout << "\nReady! Connect with:\n";
    std::cout << "  ffplay rtsp://localhost:" << port << URL_TAIL << "\n";
    std::cout << "  ffplay -rtsp_transport tcp rtsp://localhost:" << port << URL_TAIL << "\n";
    std::cout << "  ffplay -fflags nobuffer -flags low_delay -framedrop -probesize 32 -analyzeduration 0 rtsp://localhost:" << port << URL_TAIL  << "\n";

    // --- Main loop ---
    while (g_running) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // --- Stop (reverse order: upstream first) ---
    std::cout << "Stopping camera..." << std::endl;
    camera.stop();

    std::cout << "Stopping GPU OpenCV..." << std::endl;
    opencv.stop();

    std::cout << "Stopping RTSP server..." << std::endl;
    rtsp_server.stop();

    std::cout << "Done." << std::endl;
    return 0;
}
